### PR TITLE
Add StaticScatterKeyValueCache: drive mobius bias-aware external-KV (TensorScatter + nonpad_kv_seqlen) static cache

### LINKED ISSUE #2204: Data-dependent bidirectional mask (Gemma-4 vision-block) forces standard Attention over GQA — split prefill/decode decoder graphs?
### Summary

For models with a **data-dependent bidirectional attention mask** (e.g. Gemma‑4 multimodal's *vision‑block* bidirectional attention), the exported decoder is forced to use the standard ONNX `Attention` op with a float **attention bias** instead of `com.microsoft.GroupQueryAttention`. Because GQA is given up, the decoder also gives up its decode‑time fast paths — most importantly `past_present_share_buffer` (in‑place KV cache), plus XQA / Flash FastDecode.

The key observation is that **this penalty is only fundamentally required during prefill**. During token generation the bias degenerates to plain causal, so a decode‑only graph could legitimately use GQA. This looks like a strong fit for the existing **decoder pipeline** (`run_on_prompt` / `run_on_token_gen`), and I'd like guidance on whether a prefill/decode split is the recommended approach (and how to handle the KV‑cache handoff).

### Background / concrete model

- Model: **Gemma‑4‑12B** (`gemma4_unified`), exported as a 4‑model ORT‑GenAI multimodal package (vision / audio / embedding / decoder) via [mobius](https://github.com/microsoft/onnxruntime) tooling.
- Decoder: 48 layers, mixed attention — sliding‑window *local* layers (`head_dim=256`) interleaved with *global* full‑attention layers (`head_dim=512`).
- HuggingFace semantics: contiguous **image‑token** blocks attend **bidirectionally** to each other (audio/text stay causal). This is the "vision‑block overlay".

### Why GQA is dropped

The bidirectional image‑block mask cannot be expressed by `GroupQueryAttention` (causal / local‑window only — it rejects an arbitrary `attention_bias`). So mobius bakes the full mask (causal + sliding window + padding + the block‑wise bidirectional OR) into a **float additive bias** and runs the standard ONNX `Attention` op with `is_causal=0`. Result: the decoder graph is **48× `Attention`, 0× GQA**, and `genai_config.json` therefore has `past_present_share_buffer` effectively off (no in‑place KV cache).

This is **numerically correct** everywhere (verified against HF — text/image/audio all match). It's purely a performance trade‑off.

### The prefill vs decode asymmetry (the actual point)

The bidirectional bias only matters **when image tokens are queries**, which only happens during **prefill** (the image placeholders are in the current input sequence). During **decode**:

- the single query is a freshly generated **text** token,
- image tokens are only **keys** (already in past KV), never queries,
- so the new token just attends **causally** to all past KV → the overlay bias is a **no‑op**.

⇒ A decode step is plain causal attention and is fully **GQA‑eligible**. But because mobius emits a *single static graph* that must also serve prefill, **every decode step** pays the no‑GQA cost: no `past_present_share_buffer`, KV cache reallocated/copied as the sequence grows, no XQA/FastDecode. For long generations this is the dominant cost.

### Proposal / question: split prefill and decode decoder graphs?

ORT‑GenAI already appears to support exactly this via the **decoder pipeline** (`Config::Model::Decoder::PipelineModel` with `run_on_prompt` / `run_on_token_gen`, dispatched in `decoder_only_pipeline.cpp`). Concretely the idea would be:

| Stage | Graph | Mask | Op | Flags |
|-------|-------|------|----|-------|
| Prefill | `decoder_prefill.onnx` | causal + sliding + padding + **block OR** (float bias) | `Attention`, `is_causal=0` | `run_on_prompt=true`, `run_on_token_gen=false` |
| Decode | `decoder_decode.onnx` | plain causal | `GroupQueryAttention` + `past_present_share_buffer` | `run_on_prompt=false`, `run_on_token_gen=true` |

This would keep prefill correct while recovering GQA's in‑place KV cache and fast decode kernels.

**Questions for maintainers:**

1. Is the `run_on_prompt` / `run_on_token_gen` decoder pipeline the intended/supported mechanism for a heterogeneous prefill (standard `Attention`, dynamic present KV) + decode (`GroupQueryAttention`, pre‑allocated shared‑buffer KV) split?
2. **KV‑cache handoff** is the crux: prefill (`Attention`) emits a dynamic `present.*` (separate past/present), while the decode (`GQA`) graph expects a pre‑allocated `past_present_share_buffer`. Is there an existing path for the runtime to materialize the prefill KV into the GQA shared buffer (copy on first decode step), or must both stages agree on the same KV layout/allocation? Any precedent (e.g. NPU decoder‑pipeline models)?
3. Are there known examples of two pipeline stages sharing KV cache state, or is KV cache assumed private per pipeline model today?

### Alternative (ORT core, not genai)

Alternatively, if `com.microsoft.GroupQueryAttention` accepted an additive `attention_bias` (or a block/custom mask input), a **single** GQA graph could serve both prefill and decode and this whole split would be unnecessary. Filing here first because the pipeline mechanism seems to already exist; happy to mirror to onnxruntime core if that's the preferred fix.

### Environment

- onnxruntime‑genai built from source (CUDA), ORT 1.27.
- Model exported via mobius; package verified correct (text/image/audio match HF reference). This issue is strictly about decode‑time performance, not correctness.

### PR DESCRIPTION
**Pull request by titaiwangms**
## What

Adds `StaticScatterKeyValueCache`, a new KV-cache variant that lets onnxruntime-genai
drive a **mobius-exported bias-aware external-KV static-cache decoder** directly. It:

- Reuses the existing `past_present_share_buffer` lifecycle: each `key_cache.{i}` /
  `value_cache.{i}` is a pre-allocated 3D `[batch, max_seq_len, kv_hidden]` buffer that
  the graph updates **in place** via opset-24 `TensorScatter`, so past and present alias
  one `OrtValue` and never need rebinding between steps.
- Adds a small `[B]` int64 **`write_indices` / `nonpad_kv_seqlen` producer** in
  `input_ids` (driven by a standalone `StaticScatterIndexTracker`): `write_indices` is the
  scatter row offset (valid tokens before this step), `nonpad_kv_seqlen` is the valid
  cached length after this step (Attention's per-batch `seqlens_k`).
- **Factory auto-detect**: `CreateKeyValueCache` selects this variant when the model
  declares **both** `write_indices` and `nonpad_kv_seqlen` inputs (kept in lockstep with
  the producer gate so a half-declared model can't create the cache with unbound indices).
- Consumes mobius's 3D external-KV emission directly (vs the 4D
  `[batch, num_kv_heads, seq, head_dim]` default layout), and supports **per-layer-varying
  `kv_hidden`** (e.g. Gemma-4 sliding GQA vs global MQA), read from each layer's own
  declared input shape. The batch dim is taken from `BatchBeamSize()` (symbolic in the
  graph); only `max_seq_len` / `kv_hidden` must be static.

## Why

genai's existing GQA `past_present_share_buffer` path cannot carry an **arbitrary float
attention bias**. mobius now emits float-bias decoders (causal + sliding-window built
in-graph) whose KV is an external `TensorScatter` cache — that path *can* carry the bias.
This PR is the **host-side consumer** for that graph, recovering near-GQA decode behaviour
on the memory-efficient-attention (MEA) path for bias-aware models like Gemma-4.

## Relationship to onnxruntime-genai#2204

#2204 proposed **Path A** = a prefill/decode GQA graph-split. This PR implements **Path B**
= a single `TensorScatter` external-KV graph, which sidesteps the cross-stage KV handoff
that #2204 flagged as the crux. Referenced as **related**; this PR does **not** close #2204.

## Mobius side

The bias-aware external-KV graph this cache consumes is emitted by mobius
(onnxruntime/mobius#367, issue onnxruntime/mobius#366). This PR is the genai-side consumer.

## Testing

- `StaticScatterKeyValueCache.EndToEndFixtureParity` — bit-exact vs an ORT-CPU-MEA golden
  on a committed slice-A fixture (prefill seq=4 @ write 0/nonpad 4, then decode @ write
  4/nonpad 5): logits argmax/sum + `updated_key_cache.0` sums.
- `StaticScatterKeyValueCache.RewindToThrows` — deny-by-default guard: `RewindTo()` throws
  (the `write_indices`/`nonpad_kv_seqlen` stream has no rewind hook, so a silent no-op would
  desync the tracker; matches the `LFM2Cache` / `WindowedKeyValueCache` siblings).
- 5 `StaticScatterIndexTracker` producer tests pinning the off-by-one / init contract.
- **7/7 green**. CPU / MEA path — **not** gated on onnxruntime#28958.

## Scope / limits

- `batch == 1` (beam search unsupported, throws).
- MEA ≈ GQA decode ceiling, **not** Flash: the arbitrary float bias precludes the Flash path.
- Per-layer `kv_hidden` supported (Gemma-4 sliding/global mix).
- The slice-A fixture is committed under `test/models/` per the existing test-model
  convention, tracked via a `.gitignore` allowlist entry matching its siblings.

### CODE REVIEW

File: src/config.h [hunk 1]
```diff
@@ -344,6 +346,10 @@ struct Config {
         std::string past_sequence_length{Defaults::PastSequenceLengthName};
         std::string current_sequence_length{Defaults::CurrentSequenceLengthName};
         std::string total_sequence_length{Defaults::TotalSequenceLengthName};
+        // Static-scatter (TensorScatter) KV cache driver inputs ([batch] int64).
+        // Presence of write_indices selects StaticScatterKeyValueCache.
+        std::string write_indices{Defaults::WriteIndicesName};
+        std::string nonpad_kv_seqlen{Defaults::NonpadKvSeqlenName};
```
[the remark this answers is not in the document]
Comment (titaiwangms) [reply src/config.h:354]: Addressed in 7c0e90a — the comment now clarifies that a static-scatter cache is selected by the presence of **both** `write_indices` **and** `nonpad_kv_seqlen` (encapsulated as `IsStaticScatterCache()`), not `write_indices` alone. Thanks!

File: src/models/kv_cache.cpp [hunk 2]
```diff
@@ -708,6 +708,119 @@ void ModelManagedKeyValueCache::RewindTo(size_t index) {
...
+  return model.session_info_.HasInput(model.config_->model.decoder.inputs.write_indices) &&
+         model.session_info_.HasInput(model.config_->model.decoder.inputs.nonpad_kv_seqlen);
+}
+
+StaticScatterKeyValueCache::StaticScatterKeyValueCache(State& state)
+    : state_{state},
+      layer_count_{model_.config_->model.decoder.num_hidden_layers} {
+  if (state_.params_->search.num_beams != 1) {
+    throw std::runtime_error("Beam search (num_beams > 1) is not supported by the static-scatter KV cache.");
+  }
```
[the remark this answers is not in the document]
Comment (titaiwangms) [reply src/models/kv_cache.cpp:743]: Addressed in 7c0e90a — the `StaticScatterKeyValueCache` constructor now explicitly throws when `BatchBeamSize() != 1`, so `batch_size > 1` is rejected too (not just `num_beams > 1`). The index tracker advances a single slot per step, so a batch beam size of 1 is required.

File: src/models/input_ids.cpp [hunk 3]
```diff
@@ -29,6 +29,22 @@ DefaultInputIDs::DefaultInputIDs(State& state)
     *past_sequence_length_->GetTensorMutableData<int32_t>() = -1;
   }
 
+  if (model_.session_info_.HasInput(model_.config_->model.decoder.inputs.write_indices) &&
+      model_.session_info_.HasInput(model_.config_->model.decoder.inputs.nonpad_kv_seqlen)) {
+    if (state_.params_->BatchBeamSize() != 1) {
+      throw std::runtime_error("Batch size must be 1 for write_indices and nonpad_kv_seqlen inputs");
```
[the remark this answers is not in the document]
Comment (titaiwangms) [reply src/models/input_ids.cpp]: Addressed in 7c0e90a — reworded to "Batch beam size (batch_size * num_beams) must be 1" to match the `BatchBeamSize()` check. Thanks!

File: src/models/input_ids.cpp [hunk 4]
```diff
@@ -67,6 +90,16 @@ void DefaultInputIDs::Update(DeviceSpan<int32_t> new_tokens) {
     *past_sequence_length_->GetTensorMutableData<int32_t>() += new_sequence_length;
   }
 
+  if (write_indices_ && nonpad_kv_seqlen_) {
+    if (state_.params_->BatchBeamSize() != 1) {
+      throw std::runtime_error("Batch size must be 1 for write_indices and nonpad_kv_seqlen inputs");
```
[the remark this answers is not in the document]
Comment (titaiwangms) [reply src/models/input_ids.cpp]: Addressed in 7c0e90a — same reword applied here: "Batch beam size (batch_size * num_beams) must be 1", consistent with the `BatchBeamSize()` condition.

**Review by titaiwangms — COMMENTED** [review]
## Review summary — StaticScatterKeyValueCache (Path B consumer)

Reviewed at `2fd84e8` by a multi-model review team plus independent call-trace verification. **Mergeable as scoped infrastructure; no Critical correctness defect found.**

### Verified correct ✅
- **Prefill index ordering is sound** (I traced `decoder_only.cpp`): `input_ids_.Update()` → `StaticScatterIndexTracker::Advance(N)` runs **before** `State::Run`, so prefill binds `write=0, nonpad=N` — the ctor's `0/0` default is dead, and the feared one-step lag is **not present**.
- Index semantics agree with mobius: `write=valid_before`, `nonpad=valid_after`, satisfying mobius's `nonpad == write + S_q` for unpadded chunks. The tracker unit tests pin the off-by-one contract well.
- `RewindTo` fail-loud throw, factory predicate in lockstep with the producer gate (both inputs required), per-layer-varying `kv_hidden`, and batch==1 enforced on both sides — all consistent.

### Major — factory hard-fails instead of falling back
`IsStaticScatterCache()` selects on presence of **both** driver inputs *before* the other cache types in `CreateKeyValueCache`. A model that declares those inputs but uses 4D KV would hit the constructor's `rank != 3` throw rather than falling back to Default/Windowed. Name collision is unlikely (low real risk), but consider tightening the predicate (require the rank-3 static layout) or making the factory attempt-and-fall-back. (`kv_cache.cpp:711-719, 1069-1076`)

### Minor
- **Layer-index parse**: `std::stoi(idx_str)` accepts malformed names (`"0.bad"` → 0) and doesn't reject duplicates. Use `std::from_chars` / `stoi` with `pos == size` and validate non-negative + unique. (`kv_cache.cpp:740-746`)
- **Test runtime probe**: `StaticScatterRuntimeAvailable()` skips on any exception message containing `"TensorScatter"`. A real TensorScatter regression could be misclassified as "kernel absent" and silently skip parity. Match the precise missing-kernel signature (`"Could not find an implementation"` + `"TensorScatter(24)"`) and rethrow others. (`test/static_scatter_kv_cache_test.cpp`)
- **E2E parity strength**: the fixture asserts `argmax` + `sum` (1e-2 tol). `sum` can pass wrong-but-sum-preserving outputs (permutation/cancellation). Add full-tensor or slice-wise parity against the golden.

### Open question
- Decode token equal to `pad_token_id` → `Advance(0)` leaves `write_index`/`nonpad` unchanged; two such steps alias a cache slot. Mirrors the existing `current_sequence_length` logic and is benign if `pad==eos`. Confirm `pad_token_id` can't be a legal mid-stream generated token for the targeted models.

### Scope clarity (not a defect)
- **batch==1 is an enforced, documented MVP constraint** (driver tensors allocated `[1]`, throws on `BatchBeamSize()!=1`), even though the mobius graph + tests support `[batch]`. Intentional on both sides — worth tracking as a known limitation, not a break.
- KV name mapping: this consumer reads names from `genai_config.json`, while mobius emits `key_cache.{i}` / `updated_key_cache.{i}`. The fixture maps them manually; production needs the genai_config generator to emit matching names.

### Nits
- `input_index_` / `output_index_` / `layer_shapes_` are set in `Add()` but never read back — annotate clearly or remove.
- `kv_cache.h` comment "differs **only** in the allocator" erases the throw-vs-silent `RewindTo` split vs `DefaultKeyValueCache`; the struct field `write_index` (singular) vs the tensor `write_indices` (plural) is a small cross-boundary translation.

Praise: `static_scatter_indices.h` is a textbook extraction of a tricky off-by-one contract into a minimal, unit-testable type — the "crux mobius and genai must agree on" framing is exactly right. The two-tier test layout (pure-C++ tracker tests on every EP + fixture-backed e2e guarded by `#if !USE_DML` + runtime probe) is thoughtfully done.

**Review by titaiwangms — COMMENTED** [review]
## Re-review at `72e7199` — prior findings addressed; one deferred with rationale ✅

Re-checked the new commit *"Address review feedback on static-scatter KV cache"*. I independently verified the new golden by running the committed fixture `model.onnx` fresh on ORT 1.27 CPU: it matches `golden_io.npz` (≤1e-4), the scalar sum/argmax constants match, and **the C++ header arrays (`static_scatter_golden.h`) match the fresh run exactly (maxdiff = 0.0)** for all four tensors (256-elem logits ×2, 512-elem cache ×2).

| Prior finding | Status |
|---|---|
| **Minor** — `std::stoi` accepted malformed layer indices (`"0.bad"` → 0), no dedup | ✅ Factored into a model-free `DiscoverKvLayerIndices()` using `std::from_chars` with a complete-consume check (`parse_end == end`), non-negative guard, and duplicate rejection. New unit tests cover well-formed/sort, throw-on-malformed, throw-on-duplicate. |
| **Minor** — runtime probe skipped on any `"TensorScatter"` substring (could mask a real regression) | ✅ Tightened to require **both** `"Could not find an implementation for"` **and** `"TensorScatter(24)"`; rethrows anything else. |
| **Minor** — e2e parity used only `argmax` + `sum` (passes wrong-but-sum-preserving outputs) | ✅ Added full-tensor `ExpectAllClose @ 1e-3` against a generated, provenance-documented golden header (kept the sum/argmax as headline checks). Verified the golden is authentic and reproducible. |
| **Nit** — dead `input_index_` / `output_index_` / `layer_shapes_` | ✅ Removed. |
| **Nit** — "differs only in the allocator" erased the RewindTo throw-vs-reshape split | ✅ Comment rewritten to document both the 3D-vs-4D layout and the RewindTo difference. |
| **Open question** — decode token == `pad_token_id` → `Advance(0)` slot aliasing | ✅ Documented the assumption in `StaticScatterIndexTracker::Advance` (safe because `pad_token_id == eos_token_id` ends the sequence before the stalled slot is reread; would only break if a model made `pad` a legal mid-stream token). |
| **Major** — factory hard-fails (`rank != 3` throw) instead of falling back to Default/Windowed | ⏳ **Deferred as a tracked follow-up** with an in-code rationale (the rank-3 tightening touches the sensitive factory-selection path and risks layer-discovery drift for sparse/hybrid models lacking a layer-0 KV input; collision considered unlikely). Reasonable disposition — acceptable to land now and follow up. |

`DiscoverKvLayerIndices` logic reviewed by reading: the `size() > prefix+suffix` guard keeps `idx_str` non-empty, and `from_chars`/negative/duplicate checks are correct. I did not rebuild the genai C++ (expensive) but verified the golden provenance directly, which is the strongest available check. **LGTM** with the factory fallback tracked separately.
